### PR TITLE
Modular zombie spawner

### PR DESCRIPTION
**Pull request by HSGamer**
This is a big one.
Once this is done, the zombies will be split into modules that can be added and removed individually.
This also allows developers to add more custom zombies.

### CODE REVIEW

File: src/main/java/plugily/projects/villagedefense/arena/managers/spawner/FastZombieSpawner.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Override
+    public double getSpawnRate(Arena arena, int wave, int phase, int spawnAmount) {
+        if (arena.getZombies().isEmpty() || spawnAmount < 5) {
+            return 1;
+        } else if (phase == 5 && wave <= 7) {
+            return 2d / 3;
+        } else {
+            return 0;
+        }
```
Comment (montlikadani) [inline src/main/java/plugily/projects/villagedefense/arena/managers/spawner/FastZombieSpawner.java:32]:
```suggestion
        if (spawnAmount < 5 || arena.getZombies().isEmpty()) {
            return 1;
        }

        if (phase == 5 && wave <= 7) {
            return 2d / 3;
        }

        return 0;
```

File: src/main/java/plugily/projects/villagedefense/arena/managers/spawner/FastZombieSpawner.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    public int getFinalAmount(Arena arena, int wave, int phase, int spawnAmount) {
+        if (arena.getZombies().isEmpty()) {
+            return spawnAmount;
+        } else if (spawnAmount < 5) {
+            return spawnAmount;
+        } else if (phase == 5 && wave <= 7) {
+            return spawnAmount;
+        } else {
+            return 0;
+        }
```
Comment (montlikadani) [inline src/main/java/plugily/projects/villagedefense/arena/managers/spawner/FastZombieSpawner.java:45]:
```suggestion
        return (spawnAmount < 5 || (phase == 5 && wave <= 7) || arena.getZombies().isEmpty()) ? spawnAmount : 0;
```
With ternary operator is always better :)

File: src/main/java/plugily/projects/villagedefense/arena/managers/ZombieSpawnManager.java [hunk 3]
```diff
@@ -131,26 +126,22 @@ public void spawnZombies() {
...
-      for(int i = 0; i <= spawn; i++) {
-        if(arena.getOption(ArenaOption.ZOMBIES_TO_SPAWN) > 0) {
-          arena.spawnFastZombie(random);
-        }
-      }
-    }
+//    if(arena.getZombies().isEmpty()) {
+//      for(int i = 0; i <= spawn; i++) {
+//        if(arena.getOption(ArenaOption.ZOMBIES_TO_SPAWN) > 0) {
+//          arena.spawnFastZombie(random);
```
Comment (montlikadani) [inline src/main/java/plugily/projects/villagedefense/arena/managers/ZombieSpawnManager.java:132]: ...and where the fast zombies will be spawned?

Comment (HSGamer) [reply src/main/java/plugily/projects/villagedefense/arena/managers/ZombieSpawnManager.java:132]: This is just a draft. Wait for the full PR. I have just converted FastZombie, not others.

Comment (HSGamer) [reply src/main/java/plugily/projects/villagedefense/arena/managers/ZombieSpawnManager.java:132]: There will be a manager for zombie spawners. Be patient.

**Comment by HSGamer** [comment]
Ready to test and review

File: src/main/java/plugily/projects/villagedefense/arena/managers/ZombieSpawnerRegistry.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  /**
+   * Get the zombie spawner by its name
+   *
+   * @param name the name
+   * @return the zombie spawner
+   */
+  public Optional<ZombieSpawner> getSpawnerByName(String name) {
+    return zombieSpawnerList.parallelStream()
```
Comment (montlikadani) [inline src/main/java/plugily/projects/villagedefense/arena/managers/ZombieSpawnerRegistry.java:84]: `stream()` instead

**Comment by montlikadani** [comment]
Did you tested each of the zombie type, fast zombie, hard, invisible etc if it works?

File: src/main/java/plugily/projects/villagedefense/arena/managers/spawner/SimpleZombieSpawner.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+/**
+ * The interface for simple zombie spawner
+ */
+public interface SimpleZombieSpawner extends ZombieSpawner {
+  /**
+   * Get the minimum wave to spawn the zombies
+   *
+   * @return the wave
+   */
+  int getMinWave();
```
Comment (montlikadani) [inline src/main/java/plugily/projects/villagedefense/arena/managers/spawner/SimpleZombieSpawner.java:19]: I think a default method would be better here, so the minimum wave number is 1. After this you can remove from implementations where the min wave is 1.

**Comment by HSGamer** [comment]
> Did you tested each of the zombie type, fast zombie, hard, invisible etc if it works?

I've spawned all zombies. It should work for all current zombies.

**Comment by montlikadani** [comment]
Do you have any other pending changes? Otherwise I'm merging this.

**Comment by HSGamer** [comment]
> Do you have any other pending changes? Otherwise I'm merging this.

It is finished. You can merge it now.

**Pull request merged**